pkg/ride/math/math.go

```diff
@@ -64,52 +99,98 @@ func Fraction(value, numerator, denominator int64) (int64, error) {
 }
 
-func Log(base, exponent int64, baseScale, exponentScale, resultScale int, mode decimal.RoundingMode) (int64, error) {
-	if baseScale < 0 || baseScale > 8 ||
-		exponentScale < 0 || exponentScale > 8 ||
-		resultScale < 0 || resultScale > 8 {
-		return 0, errors.New("log: invalid scale")
-	}
-	b := decimal.WithContext(decimal.Context128).SetMantScale(base, baseScale)
-	e := decimal.WithContext(decimal.Context128).SetMantScale(exponent, exponentScale)
+func log(base, exponent *decimal.Big, resultScale int) (*decimal.Big, error) {
 	r := decimal.WithContext(decimal.Context128).SetMantScale(0, resultScale)
 	bl := decimal.WithContext(decimal.Context128)
-	math.Log(bl, b)
+	math.Log(bl, base)
 	if bl.Context.Err() != nil {
-		return 0, errors.New(bl.Context.Conditions.Error())
+		return nil, errors.New(bl.Context.Conditions.Error())
 	}
 	el := decimal.WithContext(decimal.Context128)
-	math.Log(el, e)
+	math.Log(el, exponent)
 	if el.Context.Err() != nil {
-		return 0, errors.New(el.Context.Conditions.Error())
+		return nil, errors.New(el.Context.Conditions.Error())
 	}
 	r.Quo(bl, el)
 	if r.Context.Err() != nil {
-		return 0, errors.New(r.Context.Conditions.Error())
+		return nil, errors.New(r.Context.Conditions.Error())
 	}
-	res, err := convertToResult(r, resultScale, mode)
+	return r, nil
+}
+
+func Log(base, exponent int64, baseScale, exponentScale, resultScale int, mode decimal.RoundingMode) (int64, error) {
+	if !checkScales(baseScale, exponentScale, resultScale) {
+		return 0, errors.New("invalid scale")
+	}
+	b := decimal.WithContext(decimal.Context128).SetMantScale(base, baseScale)
+	e := decimal.WithContext(decimal.Context128).SetMantScale(exponent, exponentScale)
+	r, err := log(b, e, resultScale)
 	if err != nil {
-		return 0, errors.Wrap(err, "log")
+		return 0, err
 	}
-	return res, nil
+	return convertToIntResult(r, resultScale, mode)
+}
+
+func LogBigInt(base, exponent *big.Int, baseScale, exponentScale, resultScale int, mode decimal.RoundingMode) (*big.Int, error) {
+	if !checkScalesBigInt(baseScale, exponentScale, resultScale) {
+		return nil, errors.New("invalid scale")
+	}
+	b := decimal.WithContext(decimal.Context128).SetBigMantScale(base, baseScale)
+	e := decimal.WithContext(decimal.Context128).SetBigMantScale(exponent, exponentScale)
+	r, err := log(b, e, resultScale)
+	if err != nil {
+		return nil, err
+	}
+	return convertToBigIntResult(r, resultScale, mode)
 }
 
-func ModDivision(x int64, y int64) int64 {
+func ModDivision(x, y int64) int64 {
 	return x - FloorDiv(x, y)*y
 }
 
-func FloorDiv(x int64, y int64) int64 {
+func FloorDiv(x, y int64) int64 {
 	r := x / y
 	// if the signs are different and modulo not zero, round down
 	if (x^y) < 0 && (r*y != x) {
 		r--
 	}
 	return r
 }
+
+func FloorDivBigInt(x, y *big.Int) *big.Int {
+	var r *big.Int
+	if x.Sign() == y.Sign() {
+		if x.Cmp(y) < 0 {
+			// abs(y-x)/2 + x
+			r = y.Sub(y, x)
+			r = r.Abs(r)
+			r = r.Div(r, big.NewInt(2))
+			r = r.Add(r, x)
+		} else {
```

Comment:
we don't need `else` here. You should make `return` just before this line in the end of `if`